### ISSUE TITLE
.NET Standard 2.0 - .deps.json modification

### ISSUE THREAD

**Issue by darvell** [Fody/Costura#358]
Issues related to this have been closed as stale but I believe it is important that this is handled.

The `dotnet build` process in .NET Core no longer produces assemblies but rather produces a single DLL with a .deps.json file that explains file dependencies. For scenarios where one might want a private library embedded where providing a user with a myget or alternative nuget package isn't feasible for either commercial reasons (e.g. secret business sauce) or just personal desire, Costura should keep  what it embedded in the .NET Standard DLL and clean up the .deps.json. On top of this, it should try and walk through the dependency tree to ensure any sub-dependencies are embedded unless the assembly has been excluded.

This is a big task and will require working closely with some build targets and understanding how these dependency trees are built, modifying them post-build and ensuring any data derived from them (e.g. nupkg files) build off the modified version instead of the original one.

I hope to keep this issue open for a bit because it's one of the last things that I personally believe Costura needs to be feature complete, especially as .NET Core + .NET Standard is the future set by MS. It's also no easy task so this feature may need to be split in to multiple steps. I'm happy to take lead/point on this if no one really wants to have to do the deep dive on this.

I hope this is in the spirit of what Costura is trying to approach, we already have Costura for .NET Core applications but nothing for .NET Standard libraries.

(Yes, I'm a patreon/contributor. I hope you don't mind this issue being more of a brain dump rather than hyper-focused milestones but it is maybe one of the last things that to me Costura needs before it's "done".)

**Comment by SimonCropp** [comment]
Would be nice if u could share a stand alone repro (as a github repository) that illustrates how the current implementation  of costura failes in the above scenario?

Also u might want to chime in on this issue https://github.com/dotnet/coreclr/issues/20287

**Comment by darvell** [comment]
Sure, will need some time to do it but I'll @ you with a standalone repository when I have an example as my current work is under NDA unfortunately.

**Comment by SimonCropp** [comment]
Thanks 

i would also assume an "real world current work" would not be ideal as it would have significant other noise. Ideally a repro should have the fewest moving pieces necessary to illustrate the problem

**Comment by swaroop-sridhar** [comment]
@darvell: The ILLinker build targets have code to rewrite deps.json files when the set of dependencies for an app changes. You may find this code useful:
https://github.com/mono/linker/blob/44199a2c62ff221f92c33cbb4ee3e35588adaf4a/corebuild/integration/ILLink.Tasks/ILLink.Tasks.targets 
CC: @sbomer  

@fadimounir If we publish a task that performs deps.json re-writing as a service, it could be useful for other projects and frameworks too. 

CC: @jeffschwMSFT

**Comment by SimonCropp** [comment]
@swaroop-sridhar @fadimounir is does seem there is a pretty significant amount of implied knowledge to know how to re-write the deps.json. with the result likely to be fragile to future changes in the msbuild pipeline.  it would be nice to see a simpler way to achieve it that is owned by MS.

**Comment by darvell** [comment]
This certainly got interesting. I still need to just build out the simple repro but now I'm not sure if we should even be doing this despite it being needed. Right now my only plan was to post-parse the deps.json format and build out a basic schema to work with it but that won't potentially work across msbuild versions.

There's a bit more to this too as well, if I merge a nuget package with costura in to the DLL, the .deps.json and the generated nupkg still include that  nupkg as a dependence. If the nupkg generation is based on what's on .deps.json then there might be a single place to approach this.

**Comment by ltrzesniewski** [comment]
The nuspec generation is based on `<PackageReference />` items. If you add the `PrivateAssets="all"` attribute on the `PackageReference`s of Fody and Costura they should no longer be listed as dependencies of the produced package.

I don't know to which extent leaving additional dependencies in the .deps.json file may produce unwanted effects, but if it does then note that *all* Fody weavers which remove their library reference are impacted.

**Comment by swaroop-sridhar** [comment]
Yes, `deps.json` is a build output and is processed at runtime. It doesn't influence packing.
The .net core host checks at startup that all dependencies listed in the `deps.json` manifest actually exist.
If an assembly or native dependency is not found, it throws as error (warning only for resource assemblies).

**Comment by ltrzesniewski** [comment]
@swaroop-sridhar sorry but... are you sure about this? I just tested a netcoreapp2.1 executable published as a self-contained app with invalid entries in deps.json and it ran just fine.
I also assume that Fody would be getting lots of bug reports if processed assemblies wouldn't work.

**Comment by ltrzesniewski** [comment]
Yeah, we have a problem here.

After running `dotnet publish`, the Fody weaver libraries listed in the .deps.json file are copied to the output directory, even though they are no longer referenced by the compiled project. If you try to remove such a weaver lib, you end up with this:

```
Error:
  An assembly specified in the application dependencies manifest (FodyIntegration.deps.json) was not found:
    package: 'Equals.Fody', version: '1.8.3'
    path: 'lib/netstandard1.0/Equals.dll'
```

I haven't tried this with Costura, but I suppose you'd end up with the embedded dependencies also in the publish output directory, which is not good. Not only it defeats the purpose of Costura, it duplicates them.

I think we should integrate the deps.json modification into Fody's `ReferenceCleaner`. I'm not sure if additional steps will be needed for Costura specifically.

**Comment by ltrzesniewski** [comment]
Actually what I said above is not actually a *huge* issue: if you add `PrivateAssets="all"` to the weaver packages, they are not listed in the .deps.json file and not copied to the output either.

It may be a good idea to update the docs to recommend `PrivateAssets="all"` in *all* cases... Unfortunately we can't just set `DevelopmentDependency` to true on weavers that remove their lib reference, as that would prevent the lib from being referenced in the first place 😢. Maybe we could forcefully set `ExcludeAssets="none"` on every weaver's package reference by hooking to the `CollectPackageReferences` target but that doesn't feel very clean... Another option would be for each weaver that wants to remove its library to add a such a hook to force `PrivateAssets="all"`, which is not very clean either but probably better than nothing.

This does not solve the issue for Costura though.

**Comment by darvell** [comment]
>I haven't tried this with Costura, but I suppose you'd end up with the embedded dependencies also in the publish output directory, which is not good. Not only it defeats the purpose of Costura, it duplicates them.

That is indeed the case. My use case is writing WPF applications but having all our business logic and some of our libraries that we give to other developers embedded as one fat DLL. Now that I've made the move to targeting only .NET Standard 2.0, distribution has become a pain point. What gets worrying even if .deps.json is ignored apparently (though it seems importing a .NET Standard 2.0 nupkg in to a .NET Framework 4.7.2 application uses that info to  work out what packages it needs), the file is still there listing all those assemblies that aren't supposed to be known to exist. It seems less about .NET Standard 2.0 because just simply using the new project system (15.9?) has changed up way too much. 

The privateassets modification doesn't really work because if I have one of those internal packages that I intend to embed depend on a public nupkg like Refit, suddenly now I don't have refit and any user consuming my package just can't use it. Maybe my personal use-case is just too odd and it'd be better off to writing a build system that does a lot of post-build analysis work to remove what's private but keep in what's okay to be public. It's just frustrating to see that the new project system has changed something that worked fine for literal years to having oddities.

It's going to take a lot of work to understand the feasibility of making fody/costura work 1:1 with how it used to for .NET Standard/new project format, maybe building a custom open sourced post-build tool for now would be best though I know Simon would love to just see this fixed and working as expected (something I still hope to do/PR depending on how much time I can find to allocate to this, it gets weirder when you have to make a business case as to why you need a couple weeks to go deep in to how the new build tools work).

>is does seem there is a pretty significant amount of implied knowledge to know how to re-write the deps.json. with the result likely to be fragile to future changes in the msbuild pipeline. it would be nice to see a simpler way to achieve it that is owned by MS.

With how much deep MS currently is in trying to get .NET Core 3 with desktop support out the door, I don't see this happening for a while even if a good handful of employees probably want it themselves.

**Comment by ltrzesniewski** [comment]
I took a quick look at Costura's source: it [removes](https://github.com/Fody/Costura/blob/39f139588f47ddf8c1a92120664261b75eb82dca/Costura.Fody/ResourceEmbedder.cs#L189) embedded assemblies from `ReferenceCopyLocalPaths`.

I think I'll take a look at the Fody weaver libraries issue this week-end, and I also think I'll remove entries from .deps.json based on what has been removed from `ReferenceCopyLocalPaths` during weaving. This will certainly require changing a few things in the way Fody caches this information. If everything works out, I *think* this Costura issue should just solve itself. No promises though ;)

**Comment by ltrzesniewski** [comment]
So there are two kinds of .deps.json files today: the one produced by `dotnet build`, and the one produced by `dotnet publish`. The first one lists all dependencies (including those with `PrivateAssets="all"`), while the second one lists only the dependencies that get really published.

Some of this is apparently going to change in .NET Core 3, due to efforts like https://github.com/dotnet/cli/issues/10061 (see https://github.com/dotnet/sdk/pull/2646). Maybe not the two different .deps.json files but at least the logic of which files end up in the build output.

Due to these moving parts, I thought it would be better to *prevent* unwanted dependencies from getting into the .deps.json file in the first place rather than removing them from there afterwards (which would also require to walk a dependency graph and remove unneeded dependencies transitively).

The .deps.json file contents are generated from `@(ReferencePath)`. Maybe Fody could remove unwanted items from there just like it does for `@(ReferenceCopyLocalPaths)`. A simple approach would be to remove any `ReferencePath` item that was removed from `ReferenceCopyLocalPaths`. But I'm worried this could cause unwanted behavior elsewhere. For instance, assemblies embedded by Costura are *still* kept as references, so maybe removing them from `ReferencePath` doesn't feel quite right. On the other hand, those have effectively vanished.

The ILLinker task mentioned earlier adds references to remove to `@(_PublishConflictPackageFiles)`. There's also `@(_ConflictPackageFiles)` which is used in both .deps.json generation targets (on build and publish). So maybe adding these there would be safer (they are only used as input to the `GenerateDepsFile` task), but those are "private" items, which a new SDK version could easily change at any time. Also, looking at the `GenerateDepsFile` tasks source code, it's not designed to work with non-NuGet references (such as project references).

**TL;DR:**

I can think of 3 ways to remove unwanted references from .deps.json:

1. Post-patch the .deps.json file: requires dependency walking, requires referencing a JSON lib in Fody
2. Remove items from `@(ReferencePath)`: may have unintended consequences
3. Remove items from `@(_PublishConflictPackageFiles)` or `@(_ConflictPackageFiles)`: changes private items, probably won't work for project references

Some of these also require preventing referenced assemblies from being copied to the publish output directory.

I'd like to avoid 1, as it can get complicated pretty fast. I'd maybe go for 3 if I only had to handle Fody weavers, but Costura may actually require 2.

Thoughts?

**Comment by darvell** [comment]
Looks like 3 has known issues from the get go so maybe not the best path. 2 is interesting but I think the only way of finding out what happens is just by doing. It's not much but I can provide some testing on large nested projects to see what side effects happen, if any.

The worrying thing about 2 is we'd probably lose nested dependencies. e.g. if my embedded library depends on newtonsoft.json, the output is going to break since now absolutely nothing knows that library was needed until execution time. Well, I think that's what would happen.

Maybe post-patching is really all that can be done to avoid a lot of problematic cases, the problem is with no schema or anything to reference, it can break easy. Is there anything that we reference in the ILLinker task to get an actual managed model of the data rather than just looking at the JSON ourselves and hoping that it doesn't change over time? If there was a DLL or something along those lines with a class for the .deps.json format, it'd be great and probably can save a lot of time/effort.

**Comment by ltrzesniewski** [comment]
> The worrying thing about 2 is we'd probably lose nested dependencies. e.g. if my embedded library depends on newtonsoft.json, the output is going to break since now absolutely nothing knows that library was needed until execution time.

Not necessarily. Dependencies have all been resolved at this point (it's one of the earliest steps in the build actually), and we'd only remove what Fody (and Costura) removes. That could work, but I won't guarantee it 😉 

There's a `ResolvePublishAssemblies` target in .NET Core SDK v2, and I'm not sure what it does exactly - it could be a problem, but probably not. It's gone in SDK v3.

> Is there anything that we reference in the ILLinker task to get an actual managed model of the data

From what I can tell ILLinker only [does what I described under number 3 above](https://github.com/mono/linker/blob/44199a2c62ff221f92c33cbb4ee3e35588adaf4a/corebuild/integration/ILLink.Tasks/ILLink.Tasks.targets#L523-L524).

> If there was a DLL or something along those lines with a class for the .deps.json format, it'd be great and probably can save a lot of time/effort.

JSON is not a real problem, the *real* PITA with this approach would be having to build a dependency graph and evaluate the full set of assemblies that need to be removed.

**Comment by ltrzesniewski** [comment]
@SimonCropp does number 2 look like a viable solution to you?

**Comment by darvell** [comment]
Think he might be busy for a bit due to the Fody version bump up. I know time ain't cheap but maybe you should just go ahead and fork/give it a go. I can spare a hand from time to time if you don't want to take it all on yourself, but you seem to have a pretty specific approach already in mind.

**Comment by ltrzesniewski** [comment]
Ok I did a very quick and dirty test just to see if it would work. It doesn't. The task which generates the .deps.json file complains it wasn't supplied the assembly reference:

> error NETSDK1007: Cannot find project info for 'C:\Dev\NuGet-Isolated\src\Costura\CosturaAssemblyToReference\CosturaAssemblyToReference.csproj'. This can indicate a missing project reference. [C:\Dev\NuGet-Isolated\
src\Costura\CosturaSample\CosturaSample.csproj]

I also tested option 3 since I already had all I needed: as I thought, it doesn't work with project references.

**Comment by Logerfo** [comment]
I'

**Comment by SimonCropp** [comment]
this project is being deprecated https://github.com/Fody/Costura/issues/442
